### DISCUSSION TITLE
Strategies for testing with Coalton?

### DISCUSSION THREAD

**Discussion by ornamentist** [coalton-lang/coalton#639] [Help!] [answered]
I'm currently creating tests for a Coalton project. I don't think there's (yet) native Coalton test frameworks so I'm using the CL framework `parachute` and writing the tests in CL:

```
(define-test "IEEE special float values equality"
  (false (coalton:coalton (num-eq? ZERO-f -INF-f  NaN-f)))
  (true  (coalton:coalton (num-eq? ZERO-f -INF-f -INF-f)))
  (false (coalton:coalton (num-eq? ZERO-f -INF-f -BIG-f)))
  ...
```

Where `num-eq?` and the values `ZERO-f` etc are Coalton functions and values. To avoid issues with the interaction of CL-level and Coalton-level data fixtures I find I'm pushing more test infrastructure into CL macros, which may not be a good strategy?

```
(defmacro special-f (value)
  `(intern (format nil "~a-F" (string ,value))))

(defmacro float-eq? (x y)
  `(coalton:coalton
     (num-eq? ZERO-f ,(special-f x) ,(special-f y))))

(define-test "IEEE special float values equality"
  (false (float-eq? NaN  NaN))
  (false (float-eq? NaN -INF))
  (false (float-eq? NaN -BIG))
  (false (float-eq? NaN -ONE))
```

What testing strategies do other Coalton developers use?

**Comment by eliaslfox** [comment, marked as answer]
I haven't gotten around to documenting it yet, but Coalton has integration with fiasco. Here's an example project https://github.com/eliaslfox/coalton-example that shows how to set it up.

**Reply by ornamentist** [reply]
Sweet!

**Reply by ornamentist** [reply]
In the ASDF file `https://github.com/eliaslfox/coalton-example/blob/main/coalton-example.asd` does the symbol `'#:coalton-example-test-fiasco` refer to the main lisp package under test?

**Reply by ornamentist** [reply]
Okay, I think I see from the `test/package.lisp` file how to setup the Fiasco references to the test package.

**Reply by stylewarning** [reply]
> Okay, I think I see from the `test/package.lisp` file how to setup the Fiasco references to the test package.

That's correct.
